### PR TITLE
feat(@formatjs/ts-transformer): support TypeScript 4.7 syntax

### LINKED ISSUE #3630: ```@formatjs/cli extract``` crashed in ts file that using new ts 4.7.2 feature "extends Constraints on infer Type Variables"
## Which package?
@formatjs/cli

## Describe the bug
error TS1005: '?' expected. after using ["extends Constraints on infer Type Variables"](https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/#extends-constraints-on-infer-type-variables)

code that crashed extraction of messages:

```
export type GetMetaType<State> = State extends {
  meta: infer Meta extends string
} 
  ? Meta 
  : never;
```

cli command that was executed:

```formatjs extract 'src/**/*.{ts,tsx,js,jsx}' --out-file build/extractedMessages/en.json --format scripts/formatter.js```

@formatjs/cli: 5.0.1
typescript: 4.7.2

### PR DESCRIPTION
**Pull request by pyrocat101**
Fixes https://github.com/formatjs/formatjs/issues/3630

Upgrades `ts-transformer` TypeScript dependency to 4.7

### CODE REVIEW

**Comment by pyrocat101** [comment]
```
        ts.Debug.assert(typeof typeReferenceDirectiveName === "string", "Non-string value passed to `ts.resolveTypeReferenceDirective`, likely by a wrapping package working with an outdated `resolveTypeReferenceDirectives` signature. This is probably not a problem in TS itself.");
                 ^
Error: Debug Failure. False expression: Non-string value passed to `ts.resolveTypeReferenceDirective`, likely by a wrapping package working with an outdated `resolveTypeReferenceDirectives` signature. This is probably not a problem in TS itself.
```

Seems like rules_nodejs has no support for the latest version of TypeScript? @longlho

**Comment by pyrocat101** [comment]
@longlho You might want to take a look at these type errors coming from TypeScript 4.7's updated type definitions.

**Comment by longlho** [comment]
Yeah I'm migrating to rules_js and then will upgrade TS

**Comment by longlho** [comment]
ill take over this PR

File: package.json [hunk 1]
```diff
@@ -125,9 +125,9 @@
     "test262-harness": "^9.0.0",
     "ts-jest": "27",
     "ts-loader": "^9.1.2",
-    "ts-node": "10",
+    "ts-node": "10.8.0",
```
Comment (redonkulus) [inline package.json]: Do you want this version locked down?

**Review by redonkulus — COMMENTED** [review]
One question

File: package.json (hunk 1 above)
Comment (longlho) [reply package.json]: fixed :)

**Comment by ghost** [comment]
Getting error TypeError: Cannot read properties of undefined (reading 'kind') after the upgrade to latest version (Using ttypescript)

**Pull request merged**